### PR TITLE
Fix new message is shown when send message

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

When the current user sends a message, an unread marker is shown.

### Fixed Issues

$ https://github.com/Expensify/App/issues/50469
PROPOSAL: https://github.com/Expensify/App/issues/50469#issuecomment-2406933488

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

1. Open any chat that is scrollable
2. Scroll to the top
3. Send a message and immediately scroll to the top*
4. Go back to the bottom and verify there is no unread marker

*an easier way is to go offline and no need to immediately scroll to the top, especially on small screen

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/98138384-aeb6-4969-b241-2fe71b10469c

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/e36c5111-65a2-4bcf-a14d-6fc9857efa79

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/5921bd4a-eb02-4d2a-a6f1-7c8d37c214ba

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/2abc73fb-b4b5-4557-839e-90b541a8a902

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/915ce3b1-d453-406d-9a14-606e4a9793c3

</details>

### CODE REVIEW

**Comment by bernhardoj** [comment]
I can't send message on android mWeb.

https://github.com/user-attachments/assets/5caca75d-7284-4691-9de0-81b64bd04e4a

**Comment by ikevin127** [comment]
@bernhardoj Thanks, this looks good and works well in all tests, except for this 1 issue:

> [!tip]
> ### Steps to reproduce
> 1. Open any chat that is scrollable.
> 2. Go offline (⌘ + D > Force offline -> ON).
> 3. Send 3 messages while offline (they will show with reduced opacity).
> 4. Scroll to the top such that the new offline sent messages cannot be seen.
> 5. Go back online (⌘ + D > Force offline -> OFF).
> 6. Once the messages are posted, while still at the top of the list, scroll slightly up and down.
> 7. Notice that the new message marker and pill are showing above the first of the 3 messages sent while offline, which shouldn't happen because the messages were sent by the same user (while offline).

https://github.com/user-attachments/assets/1ad1698f-e85d-4ec2-b87d-a35da00f9fa4

File: src/pages/home/report/ReportActionsList.tsx [hunk 1]
```diff
@@ -223,7 +226,7 @@ function ReportActionsList({
             const isNextMessageRead = !nextMessage || !isMessageUnread(nextMessage, unreadMarkerTime);
             const shouldDisplay = isCurrentMessageUnread && isNextMessageRead && !ReportActionsUtils.shouldHideNewMarker(reportAction);
             const isWithinVisibleThreshold = scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < (userActiveSince.current ?? '') : true;
-            return shouldDisplay && isWithinVisibleThreshold;
+            return shouldDisplay && isWithinVisibleThreshold && !lastActionAddedByCurrentUser.current;
```
Comment (bernhardoj) [inline src/pages/home/report/ReportActionsList.tsx:229]: Fixed by changing the check here to return false as long as `lastActionAddedByCurrentUser` exists. What happened before is, we only store the last action in `lastActionAddedByCurrentUser`. So, when we send message 1,2,3, the `lastActionAddedByCurrentUser` will be 3 and `unreadMarkerTime` is updated to 3's `created`.

When we receive the BE response, the 1 `created` value will be bigger than the optimistic 3 `created` value and previously, we only prevent the unread marker to show if `lastActionAddedByCurrentUser` is the same as the action that is being checked.

**Comment by ikevin127** [comment]
@bernhardoj Thanks, https://github.com/Expensify/App/pull/51113#discussion_r1808038884 fixed the issue I reported [here](https://github.com/Expensify/App/pull/51113#issuecomment-2424227463) for large layout devices (Web / Desktop) but it seems to still be reproducible on Android: mWeb for example (second half, the offline -> online case):

https://github.com/user-attachments/assets/e9ca5ddd-9b36-4db1-9fb4-eb0905d0a9ea

**Comment by bernhardoj** [comment]
Ah, that's because `lastActionAddedByCurrentUser` is only a ref which is cleared when you close the page. Hmm, this is really tricky.

**Comment by bernhardoj** [comment]
I have a new idea to solve it, but still fails for manually mark as unread.

**Comment by ikevin127** [comment]
@bernhardoj I don't know if it's worth fixing the Android: mWeb **only** edge case at the expense of the **manually mark as unread** functionality. I vote to leave it as is now as the only edge case is happening on Android: mWeb and only in this scenario:

> 1. Open any chat that is scrollable.
> 2. Go offline (⌘ + D > Force offline -> ON).
> 3. Send 3 messages while offline (they will show with reduced opacity).
> 4. Scroll to the top such that the new offline sent messages cannot be seen.
> 5. Go back online (⌘ + D > Force offline -> OFF).
> 6. Once the messages are posted, while still at the top of the list, scroll slightly up and down.
> 7. Notice that the new message marker and pill are showing above the first of the 3 messages sent while offline, which shouldn't happen because the messages were sent by the same user (while offline).

@MonilBhavsar What do you think ?

**Comment by ikevin127** [comment]
@bernhardoj Thanks for the detailed explanation!

Given my statement from https://github.com/Expensify/App/pull/51113#issuecomment-2430104049, looks like that would be one option in terms of compromising.

If you were to apply the solution you suggested above, what would be the drawbacks of that in terms of functionality other than the performance of array -> object ? ? Meaning if we apply that solution would:
- unread marker / pill wouldn't show up `for current user when posting and scrolling up fast (the issue)`
- unread marker / pill wouldn't show up `for current user when posting and scrolling up fast while on the same account but different session`
- the unread marker / pill wouldn't show up `for current user when posting while offline and scrolled up then going back online`
- mark the message as unread works correctly in both online / offline -> online cases

all of these scenarios pass the test ? From what I can tell currently it looks like we would have to compromise somewhere, am I wrong and we can have all these working ?

**Comment by bernhardoj** [comment]
So far it works well on all those cases.

> unread marker / pill wouldn't show up for current user when posting and scrolling up fast (the issue)


https://github.com/user-attachments/assets/cd270856-5571-4634-ad81-48cc119ed73c


> unread marker / pill wouldn't show up for current user when posting and scrolling up fast while on the same account but different session


https://github.com/user-attachments/assets/2ce494a3-beb8-4d0d-a27b-e8fb2d4d1b8a


> the unread marker / pill wouldn't show up for current user when posting while offline and scrolled up then going back online


https://github.com/user-attachments/assets/03c7d1a6-9c44-41f4-ac87-5b4da88d2212



> mark the message as unread works correctly in both online / offline -> online cases


https://github.com/user-attachments/assets/6434b633-a15e-4cea-87d1-931246a4191c

Notice the unread marker disappears when go online as I mentioned before

> 3. Go online (isNewMessage = false, isPreviouslyOptimistic = true, shouldIgnoreUnreadForCurrentUserMessage = true)
I expected in step 3, the unread marker stays, but it disappeared. In main, this also doesn't work well because of [isWithinVisibleThreshold](https://github.com/Expensify/App/blob/f251dbcdfdafc492d9a9908cbf40e4630b7b6d9d/src/pages/home/report/ReportActionsList.tsx#L225). But even if we remove isWithinVisibleThreshold, the issue still happens and it's because of the isPreviouslyOptimistic is true.

Here is the video showing the unread marker stays when removing `isWithinVisibleThreshold`.

https://github.com/user-attachments/assets/aadae5aa-d100-4a9c-8629-c05973aae8db

**Comment by ikevin127** [comment]
@bernhardoj I appreciate the tests for each case! I'm fine with the behaviour of the second to last video, meaning keeping `isWithinVisibleThreshold` but ultimately we should go with the solution that preserves the behaviour that is currently on staging - so if that means keeping or removing `isWithinVisibleThreshold` then we should go with that solution as the main goal here is to fix our issue without breaking previously working functionality.

You can proceed with the changes and I will test again before moving forward with completing the checklist and approving.

**Comment by bernhardoj** [comment]
@ikevin127 done. Please check! 🙇

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/88c122b8-acf5-44da-a681-2a9e8208828c

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/f6c2ec5c-0e7c-4d67-bb13-dae5cd096cbb

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/dccdd3fb-2c00-4eef-b0e1-e1815f73f389

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/6d186242-61d4-48c5-9a5c-2bcc7d130821

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/eb69740d-12bc-41e0-8eed-4a69abf61627

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/b70a70ad-ccb0-4546-b4de-dd8ae2b2e310

</details>

**Review by ikevin127 — APPROVED** [review]
@bernhardoj Thanks for the updates, code looks solid! 🎉

🟢 I just retested all scenarios including testing against the current staging build to ensure nothing was changed except related to our fix and all tests pass now, including on **Android: mWeb** where we had the issues with the marker not going away when returning online.

File: src/pages/home/report/ReportActionsList.tsx [hunk 2]
```diff
@@ -223,7 +236,14 @@ function ReportActionsList({
             const isNextMessageRead = !nextMessage || !isMessageUnread(nextMessage, unreadMarkerTime);
             const shouldDisplay = isCurrentMessageUnread && isNextMessageRead && !ReportActionsUtils.shouldHideNewMarker(reportAction);
             const isWithinVisibleThreshold = scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < (userActiveSince.current ?? '') : true;
-            return shouldDisplay && isWithinVisibleThreshold;
+
+            // Don't set unread marker for newly added message from the current user, only if there is no existing unread marker
+            const isFromCurrentUser = accountID === (ReportActionsUtils.isReportPreviewAction(reportAction) ? !reportAction.childLastActorAccountID : reportAction.actorAccountID);
+            const isNewMessage = !prevSortedVisibleReportActionsObjects[reportAction.reportActionID];
+            const isPreviouslyOptimistic = !!prevSortedVisibleReportActionsObjects[reportAction.reportActionID]?.isOptimisticAction && !reportAction.isOptimisticAction;
```
Comment (MonilBhavsar) [inline src/pages/home/report/ReportActionsList.tsx:243]: Could you please explain why we need this `isPreviouslyOptimistic` case

Comment (bernhardoj) [reply src/pages/home/report/ReportActionsList.tsx:247]: Added comment for it

File: src/pages/home/report/ReportActionsList.tsx [hunk 3]
```diff
@@ -197,6 +199,16 @@ function ReportActionsList({
             ),
         [sortedReportActions, isOffline],
     );
+    const lastAction = sortedVisibleReportActions.at(0);
+    const sortedVisibleReportActionsObjects: OnyxTypes.ReportActions = useMemo(
+        () =>
+            sortedVisibleReportActions.reduce((acc, action) => {
```
[the remark this answers is not in the document]
Comment (bernhardoj) [reply src/pages/home/report/ReportActionsList.tsx]: Done

**Review by MonilBhavsar — APPROVED** [review]
looks good, thank you all! 🚀

**Pull request merged**